torch/csrc/Exceptions.cpp

```diff
@@ -160,27 +162,38 @@ PyWarningHandler::~PyWarningHandler() noexcept(false) {
       // An error happened after the warning
       // Simply handle with the previous handler
       for(const auto& warning: warning_buffer_) {
-        auto source_location = warning.first;
-        const auto& msg = processErrorMsg(warning.second);
-        c10::Warning::warn(source_location, msg);
+        auto source_location = warning.source_location_;
+        const auto& msg = processErrorMsg(warning.msg_);
+        c10::Warning::warn(source_location, msg, warning.verbatim_);
       }
       warning_buffer_.clear();
     } else {
       pybind11::gil_scoped_acquire gil;
       auto result = 0;
-      for(const auto& warning: warning_buffer_) {
-        auto source_location = warning.first;
-        const auto& msg = processErrorMsg(warning.second);
+      for (const auto& warning: warning_buffer_) {
+        auto source_location = warning.source_location_;
+        const auto& msg = processErrorMsg(warning.msg_);
         if (source_location.file == nullptr) {
           result = PyErr_WarnEx(PyExc_RuntimeWarning, msg.c_str(), 1);
-        } else {
+        } else if (warning.verbatim_) {
+          // Sets the source location from the warning
+          // Note: PyErr_WarnExplicit will disregard Python's warning filter
+          // and always appear. This is in contrast to PyErr_WarnEx,
+          // which respects the warning filter.
```

Comment:
For a follow up PR, would the solution here be in the jit case to:
- Use PyErr_WarnEx and let python show the line where the jit code is called
- Use custom logic in the jit to add the original code's line inside the msg directly
- Use custom logic here to add the cpp line info?

Reply:
None of the above.

These warnings are generated by mimicking Python's warning system in the JIT. They don't trigger when we get a CPP warning while running jitted code, they trigger when the JIT emulates a Python warning, so the JIT is (correctly) mimicking the warning originating from Python like it would in eager mode.

Reply:
Ho I see.
A bit surprising but sounds fine. Maybe a note in the JIT doc would be useful?

Reply:
There is one :P 

I'll add a test case for this behavior, too, though, to clarify the distinction.